Source/UI/NewUI/PaymentForm/View/CardNumber/CardNumberView.swift

```diff
@@ -73,18 +73,27 @@ extension CardNumberView: TextFieldViewDelegate {
   func textFieldShouldReturn() -> Bool { true }
 
   func textFieldShouldEndEditing(textField: UITextField, replacementString: String) -> Bool {
+    textField.text = replacementString
+    style?.textfield.text = replacementString
     switch viewModel.validate(cardNumber: replacementString) {
     case .some(let schemeIcon):
       self.schemeIcon = schemeIcon
       updateError(show: false)
     case nil:
       updateError(show: true)
     }
-
     return true
   }
 
   func textField(_ textField: UITextField, shouldChangeCharactersIn range: NSRange, replacementString string: String) -> Bool {
+
+    // some characters are deleted
+    // notify payment view controller to disable the pay button
+    if string.count == 0 && range.length > 0 {
+      _ = viewModel.validate(cardNumber: string)
+      return true
+    }
```

Comment:
Your current approach relies on the viewModel receiving incomplete inputs (like an empty string here) and prevented from making informed decisions over changes. 

This is at odds with the other code you wrote that relies on the viewmodel receiving input for validation so the contrasting approach for the code within this method is really unhealthy for comprehension. Could you please remove this code and find a way for the viewModel to control the button state whilst integrating with existing logic and keeping it simpler?

Reply:
Also if you are returning true I can see in the code we are doing a bit of validation below to update error labels etc. Just do some test so you are not breaking anything. Its same for ExpiryDateView, SecurityCodeView & CardNumberView